phoenix-core/src/main/java/org/apache/phoenix/index/IndexMaintainer.java

```diff
@@ -883,6 +903,14 @@ public void readFields(DataInput input) throws IOException {
         dataEmptyKeyValueCF = Bytes.readByteArray(input);
         emptyKeyValueCFPtr = new ImmutableBytesPtr(Bytes.readByteArray(input));
         
+        int numIndexedExpressions = WritableUtils.readVInt(input);
```

Comment:
I added indexedColumnTypes back and used the sign of the length of emptyKeyValueCFPtr to differentiate between old and new clients. 
IndexMaintainer has changed quite a lot, If an old client is talking to a server that has been upgraded, will I have to maintain the old behavior?